storage/postgres/query_builder_test.go

```diff
@@ -117,93 +192,252 @@ var _ = Describe("Postgres Storage Query builder", func() {
+	   OR t.platform_id IS NULL);`)))
+				Expect(queryArgs).To(HaveLen(12))
+				Expect(queryArgs[0]).Should(Equal("left1"))
+				Expect(queryArgs[1]).Should(Equal("right1"))
+				Expect(queryArgs[2]).Should(Equal("left2"))
+				Expect(queryArgs[3]).Should(Equal("right2"))
+				Expect(queryArgs[4]).Should(Equal("right3"))
+				Expect(queryArgs[5]).Should(Equal("left3"))
+				Expect(queryArgs[6]).Should(Equal("right4"))
+				Expect(queryArgs[7]).Should(Equal("1"))
+				Expect(queryArgs[8]).Should(Equal("2"))
+				Expect(queryArgs[9]).Should(Equal("3"))
+				Expect(queryArgs[10]).Should(Equal("4"))
+				Expect(queryArgs[11]).Should(Equal("5"))
 			})
 		})
+	})
 
-		Context("when order by is used", func() {
-			Context("and field is uknown", func() {
-				It("should return error", func() {
-					_, err := qb.NewQuery().WithCriteria(query.OrderResultBy("unknown-field", query.AscOrder)).List(ctx, entity)
-					Expect(err).Should(HaveOccurred())
-					Expect(err.Error()).To(ContainSubstring("unsupported entity field for order by: unknown-field"))
-				})
-			})
-
-			Context("and order type is unknown", func() {
-				It("should return error", func() {
-					_, err := qb.NewQuery().WithCriteria(query.OrderResultBy("id", "unknown-order")).List(ctx, entity)
-					Expect(err).Should(HaveOccurred())
-					Expect(err.Error()).To(ContainSubstring("unsupported order type: unknown-order"))
-				})
+	Describe("Delete", func() {
+		Context("when no criteria is used", func() {
+			It("builds query to delete all entries", func() {
+				_, err := qb.NewQuery().Delete(ctx, entity)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(executedQuery).To(Equal(trim(`DELETE
+FROM visibilities USING visibilities t
+LEFT JOIN visibility_labels ON t.id = visibility_labels.visibility_id
+WHERE t.id = visibilities.id;`)))
 			})
 		})
 
-		Context("when limit is negative", func() {
-			It("should return error", func() {
-				_, err := qb.NewQuery().WithCriteria(query.LimitResultBy(-1)).List(ctx, entity)
-				Expect(err).Should(HaveOccurred())
-				Expect(err.Error()).To(ContainSubstring("limit (-1) is invalid. Limit should be positive number"))
+		Context("when label criteria is used", func() {
+			It("builds query with label criteria", func() {
+				criteria1 := query.ByLabel(query.EqualsOperator, "left1", "right1")
+				criteria2 := query.ByLabel(query.InOperator, "left2", "right2", "right3")
+				_, err := qb.NewQuery().WithCriteria(criteria1, criteria2).Delete(ctx, entity)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(executedQuery).Should(Equal(trim(`DELETE
+FROM visibilities USING visibilities t
+JOIN
+  (SELECT *
+   FROM visibility_labels
+   WHERE visibility_id IN
+	   (SELECT visibility_id
+		FROM visibility_labels
+		WHERE (visibility_labels.key = ?
+			   AND visibility_labels.val = ?)
+		  OR (visibility_labels.key = ?
+			  AND visibility_labels.val IN (?, ?)))) visibility_labels ON t.id = visibility_labels.visibility_id
+WHERE t.id = visibilities.id;`)))
```

Comment:
this is strange. I couldn't find an example on `USING` the way you're using it here. The only documentation shows it is used as an alias in `JOIN` statements.

Reply:
yeah but there is no LEFT USING (to replace LEFT JOIN) - so here im self joining first so that i can LEFT JOIN if there are not label queries